### ISSUE TITLE
Cache unfinished message

### ISSUE THREAD

**Issue by simison** [Trustroots/trustroots#148]
In case something goes wrong/session expires, browser crashes etc, save unfinished messages.

- Either on intervals (e.g. every 30seconds) or on [ng-keypress](https://docs.angularjs.org/api/ng/directive/ngKeypress)
- Can't be done with localStorage (not secure enough).
- sessionStorage might be good enough (cleared after tab closes). See [Angular-Locker](https://github.com/tymondesigns/angular-locker).
- The best would be to store this stuff at the server.
- Perhaps this is good to do after #232 (realtime messaging).

**Comment by chmac** [comment]
If we're writing to localStorage, probably makes sense to cache on every keyup, there's no network latency to consider.

**Comment by chmac** [comment]
@simison Does this really qualify as easy? :-) I thought at first, I'll just write some javascript to catch the keyup event and then save the textarea to localstorage. But it probably makes sense to do it "the angular way". What do you think?

I'll see if I can hack up something that works and I'll put it in a new branch to save polluting master if we abandon it.

**Comment by chmac** [comment]
I've added an `ng-keyup=save()` and the corresponding `save()` function. Now we need to:

- [ ] actually save the data somewhere ([angular-local-storage](https://github.com/grevory/angular-local-storage) maybe?)
- [ ] load from localStorage when returning to the thread
- [ ] clear the cache when the message is sent

**Comment by simison** [comment]
Actually, localStorage is bad for this.

1. Save unfinished message to localStorage
2. Don't send (it remains at localStorage)
3. Sign out
4. Type `localStorage` to the browser console
5. You'll see the message

Closing this, we'll need server side solution for this.

**Comment by simison** [comment]
Reusing. :-)

Updating some thoughts to the first message.

**Comment by chmac** [comment]
Not true that localStorage can't be used for this. Just add a hook to clear the message cache when a user logs out. Anyway, I'd say it's a low priority, long term goal, not an urgent problem. Also, server side means that saved messages would cross between user's sessions in different browsers.

**Comment by simison** [comment]
_There are only two hard things in Computer Science..._ ;-)

It's still not reliable; browser might just go offline, session expire etc. It's one thing to store eg. Wordpress blog post in LocalStorage (which they do) and another thing to store personal messages. Latter is a bit too sensitive to keep there.

**Comment by simison** [comment]
True about priority, moved to long term.

**Comment by chmac** [comment]
There is also sessionStorage which only persists for the duration of the session. Once the tabs have been closed, it disappears. I'm not arguing that in browser storage is definitely the right way to go here, just that I don't think security or privacy concerns are a reason to rule it out.

**Comment by simison** [comment]
Hmm, that's interesting!

[sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API) seems pretty suitable actually.

http://jsfiddle.net/uE4Pc/1/ (refresh + open new tab + refresh)

**Comment by chmac** [comment]
It's the same API without persistence, at least that's my rudimentary understanding. It's still a complicated feature though, deciding when to clear the old message if it doesn't move between browsers (and apps). If I've sent a message since then, and so on. I'd guess the best solution is to stream every keystroke to the server and create a "draft" message, which stays on the server until the user manually clears it. But something around sessionStorage might be a useful step on the road to that...
